### ISSUE TITLE
Restoring default `CTRL-X CTRL-N` behavior

### ISSUE THREAD

**Issue by rafaeln** [roxma/nvim-completion-manager#24]
This is a great plugin you wrote. Thanks for sharing!

Something I use frequently and which this plugin breaks a bit is this:

> CTRL-X CTRL-N or
> CTRL-X CTRL-P
> Further use of CTRL-X CTRL-N or CTRL-X CTRL-P will
> copy the words following the previous expansion in
> contexts unless a double CTRL-X is used.

With nvim-completion-manager, I have to type `CTRL-X CTRL-N CTRL-N` to get the same effect. Do you know any workarounds?

**Comment by roxma** [comment]
This is related to `completeopt` managed by this plugin

I don't have a neet workaround, but this might help:

```vim
inoremap <C-X><C-N> <C-X><C-N><C-N>
```

**Comment by roxma** [comment]
Maybe I could implement an auto popup completion source wiht this sort of behavior

**Comment by rafaeln** [comment]
Thanks. That's a pretty straightforward workaround. I'll give it a try. A completion source that had this sort of behavior would be even more awesome, and would be an improvement on vim's default behavior. The most awesome improvement, though, and I don't even know if this could be implemented, would be to allow this sort of ‘continue completing’ behavior with any source. You would be able, for instance, to complete a word from another tmux window, then keep copying the following words from the other tmux window with <c-x><c-n> (or whatever mapping).

**Comment by rafaeln** [comment]
Just a quick update: for some reason, `inoremap <C-X><C-N> <C-X><C-N><C-N>` doesn't do what we thought it would.

**Comment by roxma** [comment]
@rafaeln 

Hi, I've added a source for this.

Pick a key to map to `<Plug>(cm_force_refresh)`

```vim
imap <c-g> <Plug>(cm_force_refresh)
```

I also have tab mapping
```vim
      " smart tab for auto complete
      inoremap <expr> <silent> <Tab>  pumvisible()?"\<C-n>":"\<TAB>"
      inoremap <expr> <silent> <S-TAB>  pumvisible()?"\<C-p>":"\<S-TAB>"
```

`<Tab>` for selecting the popup item, `<c-g>` for refreshing the popup menu. So the key pressed in this screencast is `<TAB><c-g><TAB><c-g>...`

![output](https://cloud.githubusercontent.com/assets/4538941/23506239/34b80c10-ff82-11e6-98bd-f32f04061806.gif)

**Comment by rafaeln** [comment]
This is awesome! But it's unfortunately not working here.

I am running the latest dev version of neovim. My `.config/nvim` folder only contains the folders from `nvim-completion-manager` and the following minimal `init.vim`:

```vim
inoremap <expr> <silent> <Tab>  pumvisible()?"\<C-n>":"\<TAB>"
inoremap <expr> <silent> <S-TAB>  pumvisible()?"\<C-p>":"\<S-TAB>"
imap <c-g> <Plug>(cm_force_refresh)
```

In a buffer like the following (cursor is |)

```text
something I would like to tell you gentlemen
some|
```

I get the popup, in pink, then press `<tab>` to complete `some` into `something`, which works, and makes the popup gray, and then I press `<c-g>`, but instead of getting the expected result, the only thing that happens is that the popup menu, still only containing the word `something`, changes color back pink. Keeping pressing `<tab>` and `<c-g>` will only keep changing the popup menu from gray to pink to gray to pink and so on.

Do you have any idea of what might be going on?

**Comment by roxma** [comment]
Try press a space, then `<c-g>` to force display of popup menu.

Your first popup comes from  keyword of current buffer completion source.

The completion source added for this will be auto triggered after three matching word, then a space character.
Or a single matching word,  a space character, then ` <Plug>(cm_force_refresh)` to force update. 

It does not popup for a single word without previous matching, which does not make much sense, IMO.

**Comment by roxma** [comment]
`<Plug>(cm_force_refresh)` is not necessary for the auto triggered popup.

I've just pushed some refinement about the popup item, so it should look better.

![output](https://cloud.githubusercontent.com/assets/4538941/23511133/25cc0fac-ff97-11e6-8441-69b6c5ecb853.gif)

**Comment by rafaeln** [comment]
Oh, I see. I wasn't typing `<space>` between completing the first word and typing `<c-g>`. Thanks once more for sharing this plugin and putting together this source.

**Comment by rafaeln** [comment]
This completion source lacks one feature of vanilla `<c-x><c-n>` completion, namely, the ability to undestand other non-word characters besides `<space>`

Using vanilla vim, if my text is as follows (`|` is the cursor)

```text
\tikzset{frontier/.style=\bold}

\tik|
```

I can `<c-p><c-x><c-p><c-x><c-p><c-x><c-p><c-x><c-p>` and get

```text
\tikzset{frontier/.style=\bold}

\tikzset{frontier/.style=\bold}|
```

That is, the vanilla `<c-x><c-n>` completion mechanism understands `{`, `/`, `=` and so on as whitespace characters.

**Comment by rafaeln** [comment]
I actually suspect that NCM's multi-word completion source is an improvement on vanilla vim's `<c-x><c-p>` completion mode. Indeed, very often, when I'm using the vanilla feature to complete only the beginning of some code, after I'm done `<c-x><c-p>`'ing, I end up having the press `<c-w>` once in order to delete the last word vim added and keep only the last whitespace character it added. Using NCM's multi-word completion source, that last step won't be required.

**Comment by rafaeln** [comment]
This last commit introduced a bug, though. The picture below shows what it is.

![image](https://cloud.githubusercontent.com/assets/2940362/23588221/4bc2adfe-0188-11e7-95b3-939fab909a86.png)

**Comment by roxma** [comment]
Sorry on that, I've pushed a new fix.

**Comment by rafaeln** [comment]
This fixed it!

**Comment by roxma** [comment]
Just an update, I disabled the auto popup of this feature by default, since the current implementation scans all openning buffer.

And the auto popup after 3 previous matching word seems not appropriate for programming languages.

This setting could be overrided, `:help cm_sources_override` and the `auto_popup` option of the source.

**Comment by rafaeln** [comment]
Cool. So, talking about multi-word completion, let me report an inconsistency in behavior.

If in a line the last character before `\n` is a non-space non-word characters, such as `}`, multi-word completion wraps into the next line (which is also how vanilla vim's multi-word completion behaves). It the last characters before \n is a word character, multi-word completion is restricted to that single line.

Try it on this file:

```text
something they would like to tell you gentlemen

\tikzset{every node/.style=\it}

something they would like to tell you gentlemen}

\tikzset{every node/.style=\it}
```

**Comment by roxma** [comment]
This is duplication filtering behavior of current implementation. If I change the line order into this, the `}`  will be completed. 

```
something they would like to tell you gentlemen}

\tikzset{every node/.style=\it}

something they would like to tell you gentlemen

\tikzset{every node/.style=\it}
```

![output](https://cloud.githubusercontent.com/assets/4538941/23684403/9bb95166-03d9-11e7-898a-a19883301a67.gif)

**Comment by rafaeln** [comment]
I think I didn't explain well. Let me try again.

If you have the text (`|` is cursor)

```text
something they would like to tell you gentlemen

\tikzset{every node/.style=\it}

other things

something they would like to tell you gentlemen|
```

And you press `<c-g>`, you don't get any more suggestions, because you hit the end of the line.

But if you have the text (`|` is cursor)

```text
something they would like to tell you gentlemen

\tikzset{every node/.style=\it}

other things

\tikzset{every node/.style=\it}|
```

And you press `<c-g>`, you get `other`, from the next line down:

```text
something they would like to tell you gentlemen

\tikzset{every node/.style=\it}

other things

\tikzset{every node/.style=\it}other|
```

**Comment by roxma** [comment]
Now I understand what you mean. 

This is kind of intended when implemented this completion, becasuse I'm not sure how to refine it to a  more proper behavior. 

I need some advice.

By the way, It seems impossible for completion result to expand to the next line.

**Comment by rafaeln** [comment]
It's not really a bug. It's just inconsistent. I rarely find it useful to wrap completion into the next line, the default vim multi-word completion behavior. But sometimes it's useful. Like if you're completing into a multi-word expression, such as in the example below:

```text
Given the context of the international solidarity
movement, I would state that ...

...

The international|
```

Where `<c-x><c-p><c-x><c-p><c-x><c-p>` would give you

```text
Given the context of the international solidarity
movement, I would state that ...

...

The international solidarity movement|
```

This kind of line-wrapping completion is possible with vanilla vim multi-word completion, but it isn't possible with NCM's multi-word completion source.

**Comment by roxma** [comment]
A snippet engine  like would be able to expand multiple lines.  See this feature supported recently, https://github.com/roxma/nvim-completion-manager/issues/36

It is possible, to make the completion item expandable (snippet)

When ultisnips is triggered, it expands the whole line, then jump to the next line.


```text
something they would like to tell you gentlemen

\tikzset{every node/.style=\it}

other things

something they |
```

press `<c-g>` for completion:

```text
something they would like to tell you gentlemen

\tikzset{every node/.style=\it}

other things

something they would|
```

press ultisnips expand trigger:

```text
something they would like to tell you gentlemen

\tikzset{every node/.style=\it}

other things

something they would like to tell you gentlemen
|
```

**Comment by rafaeln** [comment]
I wasn't aware ultisnips was that flexible. I'll go read the documentation. I don't think the inconsistency is a real issue. There's always the workaround of resorting to vim's vanilla multiword completion mechanism whenever the rare occasion arises where I'd like to complete to wrap into the next line.
